Neos.Utility.Arrays/Classes/Arrays.php

```diff
@@ -218,6 +219,34 @@ public static function getValueByPath(array &$array, $path)
         }
     }
 
+    /**
+     * Returns a type safe accessor for a value in a nested array by following the specifed path.
+     *
+     * @param array $array The array to traverse as a reference
+     * @param array|string $path The path to follow. Either a simple array of keys or a string in the format 'foo.bar.baz'
+     * @param string|null $pathinfoSoFar
+     * @return ValueAccessor
+     */
+    public static function getAccessorByPath(array $array, array|string $path, string $pathinfoSoFar = null): ValueAccessor
+    {
+        if (is_string($path)) {
+            $path = explode('.', $path);
+        }
+        $key = array_shift($path);
+        $pathinfo = is_string($pathinfoSoFar) ? $pathinfoSoFar . '.' . $key : $key;
+        if (isset($array[$key])) {
+            if (count($path) > 0) {
+                return is_array($array[$key])
+                    ? self::getAccessorByPath($array[$key], $path, $pathinfo)
+                    : new ValueAccessor(null, $pathinfo);
+            } else {
+                return new ValueAccessor($array[$key], $pathinfo);
+            }
+        } else {
+            return new ValueAccessor(null, $pathinfo);
+        }
+    }
```

Comment:
There are a few possibilities for early returns that reduce cognitive load:

```suggestion
     * @param string|null $pathinfoSoFar
     * @return ValueAccessor
     */
    public static function getAccessorByPath(array $array, array|string $path, string $pathinfoSoFar = null): ValueAccessor
    {
        $pathSegments = is_string($path) ? explode('.', $path) : $path;
        $pathSegment = array_shift($pathSegments);
        $pathinfo = is_string($pathinfoSoFar) ? $pathinfoSoFar . '.' . $pathSegment : $pathSegment;
        if (!isset($array[$pathSegment])) {
            return new ValueAccessor(null, $pathinfo);
        }
        if ($pathSegments === []) {
            return new ValueAccessor($array[$pathSegment], $pathinfo);
        }
        return is_array($array[$pathSegment])
            ? self::getAccessorByPath($array[$pathSegment], $pathSegments, $pathinfo)
            : new ValueAccessor(null, $pathinfo);
    }
```

But I also struggle a bit with the last argument that has no doc comments and should probably only be used internally for the recursion AFAIS.
What about turning this into a non-recursive function like:

```suggestion
     * @return ValueAccessor
     */
    public static function getAccessorByPath(array $array, array|string $path, string $pathinfoSoFar = null): ValueAccessor
    {
        $pathSegments = is_string($path) ? explode('.', $path) : $path;
        $value = $array;
        $processedSegments = [];
        do {
            $pathSegment = array_shift($pathSegments);
            if (!array_key_exists($pathSegment, $value)) {
                $value = null;
                break;
            }
            $processedSegments[] = $pathSegment;
            $value = $value[$pathSegment];
            if (!is_array($value)) {
                if ($pathSegments !== []) {
                    $value = null;
                }
                break;
            }
        } while ($pathSegments !== []);
        return new ValueAccessor($value, implode('.', $processedSegments));
    }
```


Alternatively, we could consider reworking the existing `getValueByPath()` such that we could reuse it here:

```php
public static function getValueByPath(array $array, array|string $path, string &$processedPath = null): mixed
{
    $pathSegments = is_string($path) ? explode('.', $path) : $path;
    $value = $array;
    $processedSegments = [];
    do {
        $pathSegment = array_shift($pathSegments);
        if (!array_key_exists($pathSegment, $value)) {
            $value = null;
            break;
        }
        $processedSegments[] = $pathSegment;
        $value = $value[$pathSegment];
        if (!is_array($value)) {
            if ($pathSegments !== []) {
                $value = null;
            }
            break;
        }
    } while ($pathSegments !== []);
    $processedPath = implode('.', $processedSegments);
    return $value;
}
```

With that in place we could simplify the new method to just
```php
public static function getAccessorByPath(array $array, array|string $path): ValueAccessor
{
    $value = self::getValueByPath($array, $path, $processedPath);
    return new ValueAccessor($value, $processedPath);
}
```

(applying the last suggestion leads to one failing test because `getValueByPath()` used to throw an `InvalidArgumentException` instead of a `TypeError` – but that behavior should be unified anyways, IMO)

Reply:
Adjusted to use getValueByPath internally which used your early return suggestion aswell.